src/acquisition/flusurv/flusurv.py

```diff
@@ -142,78 +173,192 @@ def mmwrid_to_epiweek(mmwrid):
-                data_out[epiweek] = [None] * 9
-            prev_rate = data_out[epiweek][age_index]
-            if prev_rate is None:
-                # this is the first time to see a rate for this epiweek/age
-                data_out[epiweek][age_index] = rate
-            elif prev_rate != rate:
-                # a different rate was already found for this epiweek/age
-                format_args = (epiweek, obj["age"], prev_rate, rate)
-                print("warning: %d %d %f != %f" % format_args)
-
-    # sanity check the result
-    if len(data_out) == 0:
+    # Sanity check the input. We expect to see some epiweeks
+    if len(data) == 0:
         raise Exception("no data found")
 
-    # print the result and return flu data
-    print(f"found data for {len(data_out)} weeks")
+    id_label_map = make_id_label_map(metadata)
+    id_season_map = make_id_season_map(metadata)
+
+    # Create output object
+    # First layer of keys is epiweeks. Second layer of keys is groups
+    #  (by id, not age in years, sex abbr, etc).
+    #
+    # If a top-level key doesn't already exist, create a new empty dict.
+    # If a secondary key doesn't already exist, create a new key with a
+    #  default value of None if not provided.
+    data_out = defaultdict(lambda: defaultdict(lambda: None))
+
+    # data["default_data"] is a list of dictionaries, with the format
+    #     [
+    #         {'networkid': 1, 'catchmentid': 22, 'seasonid': 49, 'ageid': 0, 'sexid': 0, 'raceid': 1, 'rate': 4.3, 'weeklyrate': 1.7, 'mmwrid': 2493},
+    #         {'networkid': 1, 'catchmentid': 22, 'seasonid': 49, 'ageid': 0, 'sexid': 0, 'raceid': 1, 'rate': 20.3, 'weeklyrate': 0.1, 'mmwrid': 2513},
+    #         {'networkid': 1, 'catchmentid': 22, 'seasonid': 49, 'ageid': 0, 'sexid': 0, 'raceid': 1, 'rate': 20.6, 'weeklyrate': 0.1, 'mmwrid': 2516},
+    #         ...
+    #     ]
+    for obs in data:
+        epiweek = mmwrid_to_epiweek(obs["mmwrid"])
+        season = id_season_map[obs["seasonid"]]
+        groupname = groupids_to_name(
+            ageid = obs["ageid"], sexid = obs["sexid"], raceid = obs["raceid"],
+            id_label_map = id_label_map
+        )
+
+        # Set season description. This will be overwritten every iteration,
+        #  but should always have the same value per epiweek group.
+        data_out[epiweek]["season"] = season
+
+        rate = obs["weeklyrate"]
+        prev_rate = data_out[epiweek][groupname]
+        if prev_rate is None:
+            # This is the first time to see a rate for this epiweek-group
+            #  combo
+            data_out[epiweek][groupname] = rate
+        elif prev_rate != rate:
+            # Skip and warn; a different rate was already found for this
+            # epiweek-group combo
+            warn((f"warning: Multiple rates seen for {epiweek} "
+                   f"{groupname}, but previous value {prev_rate} does not "
+                   f"equal new value {rate}. Using the first value."))
```

Comment:
im guessing youve seen this in real data?  does it happen often and/or with large differences in value?  would it be worth halting acquisition when it happens, or maybe coming up with a way of escalating the message to where someone is sure to see it?

additionally, how would you feel about using [our logger module](https://github.com/cmu-delphi/delphi-epidata/blob/dev/src/common/logger.py) instead of the mixed `print()` and `warn()` statements?

Reply:
I reused logic that was already in the script; I haven't actually seen this in real data.

This was apparently set up to [match the behavior of the dashboard](https://github.com/cmu-delphi/delphi-epidata/blob/9f55c8197f7eea3e41e483c175e7cb87e0655e53/src/acquisition/flusurv/flusurv.py#L11-L15) that we scrape. Apparently we used to think the [data was versioned](https://github.com/cmu-delphi/delphi-epidata/blob/14c102a38c1a8d575b877f402fae676c9b7ea8ea/src/acquisition/flusurv/flusurv.py#L11-L17) or at least treat it like it was. It's unclear how much of this is sleuthing/speculation on our part, since the API we're using is a CDC internal API with no documentation. And the dashboard contains extra processing itself.

So... I could understand keeping this as-is, since the behavior matches the source's. It also would make sense for this to fail if we see contradictory values so we can investigate. Theoretically we have logs from all these runs that would tell us how often this happens, but without the script failing no one's going to look into it.

Reply:
It makes sense to use more formal logging, but given this PR is already pretty large, maybe that should be a separate change. Unclear ATM how extensively I should refactor this (metadata class, e.g.), so I'll open an issue for now.

Reply:
Logging changes proposed in https://github.com/cmu-delphi/delphi-epidata/issues/1294

Reply:
Discussion about behavior when we see two values for the same geo-date https://github.com/cmu-delphi/delphi-epidata/issues/1331